MapboxNavigation/EndOfRouteViewController.swift

```diff
@@ -0,0 +1,193 @@
+        commentView.resignFirstResponder()
+        let fireTime = DispatchTime.now() + 0.3 //Not ideal, but works for now
+        DispatchQueue.main.asyncAfter(deadline: fireTime, execute: dismissal)
+    }
+    
+    private func showComments(animated: Bool = true) {
+        showCommentView.isActive = true
+        hideCommentView.isActive = false
+        ratingCommentsSpacing.constant = ConstraintSpacing.closer.rawValue
+        preferredContentSize.height = height(for: .commentShowing)
+
+        let animate = {
+            self.view.layoutIfNeeded()
+            self.commentViewContainer.alpha = 1.0
+            self.labelContainer.alpha = 0.0
+        }
+        
+        let completion: (Bool) -> Void = { _ in self.labelContainer.isHidden = true}
+        let noAnimate = { animate() ; completion(true) }
+        animated ? UIView.animate(withDuration: 0.3, animations: animate, completion: nil) : noAnimate()
+    }
+    
+    private func hideComments(animated: Bool = true) {
+        labelContainer.isHidden = false
+        showCommentView.isActive = false
+        hideCommentView.isActive = true
+        ratingCommentsSpacing.constant = ConstraintSpacing.further.rawValue
+        preferredContentSize.height = height(for: .normal)
+        
+        let animate = {
+            self.view.layoutIfNeeded()
+            self.commentViewContainer.alpha = 0.0
+            self.labelContainer.alpha = 1.0
+        }
+        
+        let completion: (Bool) -> Void = { _ in self.commentViewContainer.isHidden = true }
+        let noAnimation = { animate(); completion(true)}
+        animated ? UIView.animate(withDuration: 0.3, animations: animate, completion: nil) : noAnimation()
+    }
+    
+    private func height(for height: ContainerHeight) -> CGFloat {
+        guard #available(iOS 11.0, *) else { return height.rawValue }
+        let window = UIApplication.shared.keyWindow
+        let bottomMargin = window!.safeAreaInsets.bottom
+        return height.rawValue + bottomMargin
+    }
+    
+    private func updateInterface() {
+        primary.text = string(for: destination)
+    }
+
+    private func clearInterface() {
+        primary.text = nil
+        stars.rating = 0
+    }
+    
+    private func string(for destination: Waypoint?) -> String {
+        guard let destination = destination else { return "Unknown" }
+        guard destination.name?.isEmpty ?? false else { return destination.name! }
+        let coord = destination.coordinate
+        return String(format: "%.2f", coord.latitude) + "," + String(format: "%.2f", coord.longitude)
+    }
+    
+    private func setPlaceholderText() {
+        commentView.text = placeholder
+    }
+}
+
+//MARK: - UITextViewDelegate
+extension EndOfRouteViewController: UITextViewDelegate {
+    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+        guard text == "\n" else { return true }
```

Comment:
Nit: could be improved with `CharacterSet.newlines`

Reply:
- [ ] use sugar